app/utils/context_runner.rb

```diff
@@ -132,6 +135,8 @@ def next_mw(ctx)
       [new_ctx, mw, :enter]
     else
       mw = new_ctx[:leave_stack].shift
+      new_ctx[:complete_stack].unshift(mw)
```

Comment:
Is there any way we could implement the desired behaviour without needing to add a new stack? My objection is that this:
* Adds complexity by adding one more moving part to be handled
* Is not obvious; the name complete_stack is not nearly as intuitive as enter_queue and leave_stack are which correspond to execution phases and order
* This is the only case where it has been needed so far (and I'm thinking if it's really needed here?) so it feels like adding this is only to support one specific use case and not a generally useful concept
* Pedestal doesn't have this

How about an approach where the retry middleware makes a copy of the remaining enter_queue at enter phase and stores it to ctx. And then in leave phase copies back the enter_queue to trigger a retry. Would this work?